### ISSUE TITLE
Weaver swallowing fatal exceptions?

### ISSUE THREAD

**Issue by armanbilge** [disneystreaming/weaver-test#503]
I originally encountered this as a [UndefinedBehaviorError](https://github.com/scala-js/scala-js/blob/0708917912938714d52be1426364f78a3d1fd269/linker-private-library/src/main/scala/org/scalajs/linker/runtime/UndefinedBehaviorError.scala#L24) on Scala.js.

```scala
//> using scala "3.1.1"
//> using platform "scala-js"
//> using jsVersion "1.9.0"
//> using lib "com.disneystreaming::weaver-cats::0.7.11"
//> using testFramework "weaver.framework.CatsEffect"

import cats.effect._
import weaver._

object MySuite extends SimpleIOSuite {
  test("boom") {
    IO(throw new VirtualMachineError {})
  }
}
```

```
$ scala-cli test bug.test.scala 
MySuite
- boom 19ms
*************FAILURES**************
MySuite
- boom 19ms
  JavaScriptException: TypeError: Cannot read properties of null (reading 'run__Lcats_data_Validated')
```

Running on the JVM completely hangs.
```scala
//> using scala "3.1.1"
//> using lib "com.disneystreaming::weaver-cats::0.7.11"
//> using testFramework "weaver.framework.CatsEffect"

import cats.effect._
import weaver._

object MySuite extends SimpleIOSuite {
  test("boom") {
    IO(throw new VirtualMachineError {})
  }
}
```

**Comment by keynmol** [comment]
I believe https://github.com/typelevel/cats-effect/issues/726 and https://github.com/disneystreaming/weaver-test/issues/222#issuecomment-802369177 are the relevant context for this.

**Comment by Baccata** [comment]
I do wonder if in some cases we could benefit form a more graceful degradation, like `IOApp`, in the CE3 case anyway. The problem is that we don't quite control the "end of the world" (the build tool does). 

We are going to be removing CE2 support, which might help find a solution.

**Comment by Baccata** [comment]
I'll try to book some of Daniel's time during work hours, see if he has thoughts on the matter.

**Comment by armanbilge** [comment]
> The problem is that we don't quite control the "end of the world" (the build tool does).

That's fair, except it almost seems to me that the swallowing is happening before that point? Since the code attempts to proceed as normal.

```
TypeError: Cannot read properties of null (reading 'run__Lcats_data_Validated')
```

When I first encountered this it was extremely confusing because it was manifesting as a NPE over here actually, as part of a `Stream#compile#foldMonoid` (I think ...):
https://github.com/disneystreaming/weaver-test/blob/d3c11c2446b6af79c7482d0a35d11268bb21d410/modules/core/src/weaver/Expectations.scala#L12-L13

All said, could be something weird with `IO` and Scala.js. Which I suppose makes it my fault 😆

**Comment by Baccata** [comment]
Could it be this ? 

https://github.com/disneystreaming/weaver-test/blob/master/modules/core/src/weaver/Test.scala#L24 

My assumption/intuition is that fatals are not passed over to the lambda argument in `handleError`.

**Comment by armanbilge** [comment]
`IO` is only supposed to capture non-fatal errors, so assuming that `F` is `IO` then your assumption is correct.

**Comment by Baccata** [comment]
Yeah so that null is definitely off, because the Validated should not even be instantiated. But then again, I have no idea how fatals are behaving on JS, so weird behaviour is not a total surprise. 

That being said it is still better than the hanging behaviour on JVM 😓

**Comment by armanbilge** [comment]
Yeah, super strange. Because this program appears to do the right thing and let the fatal bubble up.

```scala
//> using scala "3.1.1"
//> using platform "scala-js"
//> using jsVersion "1.9.0"
//> using lib "org.typelevel::cats-effect::3.3.9"

import cats.effect._

object MySuite extends IOApp.Simple {
  def run = IO(throw new VirtualMachineError {}).attempt *> IO.println("boom")
}
```

**Comment by Baccata** [comment]
Here's where most of the heavy lifting happens, if you want to have a look  : 

* JVM runner: https://github.com/disneystreaming/weaver-test/blob/master/modules/framework/src-jvm/RunnerCompat.scala
* JS runner : https://github.com/disneystreaming/weaver-test/blob/master/modules/framework/src-js/RunnerCompat.scala

The two have vastly different semantics, mostly driven by the fact that we support cross-suites resources on JVM and therefore have to weave the suites together in a single IO.

**Comment by armanbilge** [comment]
OTOH, this suite passes! 🤯 

```scala
//> using scala "3.1.1"
//> using platform "scala-js"
//> using jsVersion "1.9.0"
//> using lib "com.disneystreaming::weaver-cats::0.7.11"
//> using testFramework "weaver.framework.CatsEffect"
//> using lib "org.typelevel::cats-effect::3.3.9"

import cats.effect._
import weaver._

object MySuite extends SimpleIOSuite {
  test("boom") {
    IO(throw new VirtualMachineError {}).attempt.as(success) <* IO.println("BOOM")
  }
}
```

```
MySuite
BOOM
+ boom 3ms
```

**Comment by armanbilge** [comment]
Aha, this appears to be a Scala.js thing. Took munit for a spin and it seems to have the same issue.

```scala
//> using scala "3.1.1"
//> using platform "scala-js"
//> using jsVersion "1.9.0"
//> using lib "org.scalameta::munit::0.7.29"
//> using lib "org.typelevel::cats-effect::3.3.9"
//> using testFramework "munit.Framework"

import cats.effect._
import cats.effect.unsafe.implicits._
import munit._

class MySuite extends FunSuite {
  test("boom") {
    val io = IO(throw new VirtualMachineError {}).attempt <* IO.println("BOOM")
    io.unsafeToFuture()
  }
}
```

```
MySuite:
==> X MySuite.boom 0.03s MySuite$$anon$1: null
    at MySuite.$anonfun$1(/tmp/main6742797921247886616.js:63975)
    at <jscode>.{anonymous}()(/tmp/main6742797921247886616.js:63979)
    at scala.scalajs.runtime.AnonFunction0.apply(/tmp/main6742797921247886616.js:37703)
    at cats.effect.IOFiber.runLoop(/tmp/main6742797921247886616.js:53507)
    at cats.effect.IOFiber.execR(/tmp/main6742797921247886616.js:54328)
    at cats.effect.IOFiber.run(/tmp/main6742797921247886616.js:54770)
    at cats.effect.unsafe.FiberAwareExecutionContext.cats$effect$unsafe$FiberAwareExecutionContext$$_$execute$$anonfun$1(/tmp/main6742797921247886616.js:21726)
    at cats.effect.unsafe.FiberAwareExecutionContext$$anon$1.run(/tmp/main6742797921247886616.js:21764)
    at org.scalajs.macrotaskexecutor.MacrotaskExecutor$.execute$$anonfun$1(/tmp/main6742797921247886616.js:41914)
    at org.scalajs.macrotaskexecutor.MacrotaskExecutor$.execute$$anonfun$adapted$1(/tmp/main6742797921247886616.js:41917)
BOOM
/tmp/main6742797921247886616.js:18103
    throw $ct_jl_IllegalStateException__T__(new $c_jl_IllegalStateException(), "Promise already completed.")
    ^
```

Feel free to close unless you want to track the JVM thing.

**Comment by armanbilge** [comment]
> Took munit for a spin and it seems to have the same issue.

Well, I take that back. At least with munit the error does appear to surface, whereas weaver is completely able to swallow it.

It's still weird that the `IO.println("BOOM")` runs though in munit, which is why I think something else is wrong.

**Comment by Baccata** [comment]
Cool. I know it's a lot to ask, but any chance you could have a look at the JS runner sometime, see if you can pinpoint the issue ? (I am very much not versed into the intricacies of the JS runtime). 

In terms of workflow, the easiest way to experiment with stuff is to clone the repo and put some suites in an `src-js` folder at [this](https://github.com/disneystreaming/weaver-test/tree/master/modules/framework/cats/test) level, and run `catsJS/testOnly ...` within SBT (weaver is tested with weaver)

**Comment by Baccata** [comment]
I should add that as a tagline : `weaver-test, finding bugs in cats-effect since 1999`

**Comment by kubukoz** [comment]
Is there a resolution for the JVM hang? I've been seeing this quite a lot recently due to binary incompatibilities in upstream libraries (Smithy...) and was about to report it in weaver. Should I open a new issue?

**Comment by Baccata** [comment]
@kubukoz yes, but please accompany it with a repro 🙏
